src/environment.cpp

```diff
@@ -715,41 +694,92 @@ typedef struct {
-			auto & active_object_count_wider = abm.active_object_count_wider;
-			auto & neighbor = abm.neighbor;
+		ServerMap *map = &m_env->getServerMap();
 
+		float dtime = 0;
+		if (m_abm_timestamp) {
+			dtime = time - m_abm_timestamp;
+		} else {
+			u32 ts = getActualTimestamp();
+			if (ts)
+				dtime = time - ts;
+			else
+				dtime = 1;
+		}
+		//infostream<<"MapBlock::abm_triggers_run p="<<getPos()<<" abm_triggers="<<abm_triggers<<" size()="<<abm_triggers->size()<<" time="<<time<<" dtime="<<dtime<<" activate="<<activate<<std::endl;
+		m_abm_timestamp = time;
+		//for (const auto & abm_trigger : *abm_triggers) {
+		for (auto ir = abm_triggers->begin(); ir != abm_triggers->end() ; ++ir) {
+			ScopeProfiler sp2(g_profiler, "ABM trigger nodes test", SPT_ADD);
+			auto & abm_trigger = *ir;
+			auto & i = abm_trigger.i;
+			float intervals = dtime / i->abmws->interval;
+			int chance = (i->abmws->chance / intervals);
+			auto rnd = myrand();
+			//infostream<<"TST: dtime="<<dtime<<" Achance="<<i->abmws->chance<<" Ainterval="<<i->abmws->interval<< " Rchance="<<chance<<" Rintervals="<<intervals  <<" rnd="<<rnd <<std::endl;
+
+			if(chance && rnd % chance)
+					continue;
+
+			//infostream<<"HIT! dtime="<<dtime<<" Achance="<<i->abmws->chance<<" Ainterval="<<i->abmws->interval<< " Rchance="<<chance<<" Rintervals="<<intervals  <<" rnd="<<rnd <<std::endl;
+
+			auto n = map->getNodeTry(abm_trigger.p);
+			if (n.getContent() != abm_trigger.c) {
+				if (n)
+					ir = abm_triggers->erase(ir);
+				continue;
+			}
 			//TODO: async call for c++ abms
+			ScopeProfiler sp3(g_profiler, "ABM trigger nodes call", SPT_ADD);
 
-				i->abm->trigger(m_env, p, n,
-						active_object_count, active_object_count_wider, neighbor, activate);
+				i->abmws->abm->trigger(m_env, abm_trigger.p, n,
+						abm_trigger.active_object_count, abm_trigger.active_object_count_wider, map->getNodeTry(abm_trigger.neighbor_pos), activate);
 
 				// Count surrounding objects again if the abms added any
+/*
 				if(m_env->m_added_objects > 0) {
 					active_object_count = countObjects(block, map, active_object_count_wider);
 					m_env->m_added_objects = 0;
 				}
-
-			trigger_list.pop_front();
+*/
+		}
+		if (abm_triggers->empty()){
+			delete abm_triggers;
+			abm_triggers = nullptr;
 		}
+}
+
+void ServerEnvironment::analyzeBlock(MapBlock * block) {
 
+	u32 block_timestamp = block->getActualTimestamp();
+	if (block->m_analyzed_timestamp > block_timestamp) {
+		//infostream<<"not anlalyzing: ats="<<block->m_analyzed_timestamp<< " bts="<<  block_timestamp<<std::endl;
+		return;
 	}
-/*
-};
-*/
+	ScopeProfiler sp(g_profiler, "ABM analyze", SPT_ADD);
+	block->analyzeContent();
+	//infostream<<"ServerEnvironment::analyzeBlock p="<<block->getPos()<< " tdiff="<<block->m_changed_timestamp - block->m_analyzed_timestamp   <<" co="<<block->content_only<<std::endl;
+	m_abmhandler->apply(block);
+	block->m_analyzed_timestamp = block_timestamp + 5;
```

Comment:
do not re-analyze block more often than 5s
more value (10s ? ) = less cpu load (especially near flowing water or someshing changing)
less value (1s ?) = faster start trigger freshly placed block